### ISSUE TITLE
Checkpointing best practices

### ISSUE THREAD

**Issue by DrDanRyan** [automl/HpBandSter#42]
I think it would be really helpful to have a page describing best practices for model/result checkpointing to avoid recomputing early parts of the training when a configuration is being computed at a higher budget. It was not obvious to me that this is something that should probably be done until after reading the Hyperband paper more carefully and fully understanding the algorithm.

For example:
- If budget is related to epochs of training for a neural network, it would be good to keep checkpoints of models that have been trained up to the highest budget requested of that config so far.
- If budget is related to number of cross-validation folds, it would be good to keep track of average validation score for folds already computed at highest budget requested of that config so far.

However, I am not sure the best way to accomplish this in a multi-node setup (reading/writing pickled objects over the network to a central storage location?) Also, it seems like the SuccesiveHalving subroutine would be able to identify when a configuration has been eliminated from further consideration (won't be asked to evaluate at a higher budget) at which point the checkpoint could be deleted.

I think ultimately it would be really nice if the `Worker.compute` method received checkpoint data (if available) as part of its input and the `Worker` returns checkpoint data as part of its return value on the `compute` method and the rest (storing highest budget relevant checkpoints, deleting unneeded checkpoints) was handled by `Master`. The `Worker` would just be responsible for taking the checkpoint data and using it to warm start it's compute process for the higher budget and then returning updated checkpoint data after computing the higher budget (if less than `max_budget`).

**Comment by sfalkner** [comment]
I fully agree that it would be very nice to have an example on how to resume inside a worker. Making this part of HpBandSter in particular the `Worker.compute` method would be nice, but will probably not happen. Here are the reasons why:

- you can already achieve that behavior with minimal code (more on that later)
- the library Pyro4 used to communicate over the network only really supports build in python types. This is why you can't store numpy arrays in the config_info for example. So, sending binary data around is not really an option.
- Even if we could send binary data over the network, it would lead to an unpredictable workload on the master which would have to store all of that data somewhere.

I know these aren't really deal breaking reasons, but please keep in mind, that this library is more or less research code, and not a professionally maintained library written by a team of software engineers. 
Here is how you can actually restart from previous checkpoints assuming that all workers have access to a shared file system. The compute method gets an argument called `working_directory` which is a unique directory for the whole run. If your worker creates a subdirectory in it that can be identified with the config_id, you can store any data you want in it. If the worker already finds this directory, it can simply load the data and continue training. This works (used that myself), but requires you to store every single model, which could require a lot of disk space.

The only thing missing that you already pointed out, is a way of cleaning up while the run is still in progress. I am not sure what the best way of incorporating that into the library would be, but I could see another method for the worker, maybe called `cleanup`, that could be called every time a configuration will no longer be evaluated on a higher budget. This could then be called from within the Master automatically. I can't promise anything, but I will try to look into it. 

Side notes:
- the actual overhead from retraining is smaller than most people think. If you use the default values eta=3, you train for at most 50% more for the configurations that are evaluated on the full budget.
- if you are training neural networks with a learning rate schedule, it is more beneficial to adapt the schedule based on the budget rather than continuing. For example, if you use cosine annealing, make the schedule always end on the given budget. This improves comparability across different configurations. As I usually train with a schedule, I follow this strategy as it proved more effective for me.

Hope that helps.

**Comment by DrDanRyan** [comment]
Thanks for the response, especially the comment about eta=3 => only 50% extra compute for runs with full budget. I did some further calculations and for `eta=3, s_max=3` the total wasted compute is around 15% which makes me think it is probably not worth worrying about too much.
